File: modules/app-updater/android/src/main/java/org/tinycld/appupdater/AppUpdaterModule.kt
package org.tinycld.appupdater

import android.content.Context
import com.facebook.react.ReactApplication
import com.facebook.react.bridge.UiThreadUtil
import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition
import java.io.File
import java.security.MessageDigest
import org.json.JSONObject

class AppUpdaterModule : Module() {
    private val context: Context
        get() = appContext.reactContext ?: throw IllegalStateException("no react context")

    override fun definition() = ModuleDefinition {
        Name("AppUpdaterModule")

        Function("getEmbeddedId") { Store(context).embeddedId() }
        Function("getRuntimeVersion") { Store(context).embeddedRuntime() }
        Function("getCurrentBundleId") { Store(context).currentId() ?: Store(context).embeddedId() }
        Function("getCurrentBundleHash") {
            val store = Store(context)
            store.currentBundleHash(store.embeddedId())
        }

        AsyncFunction("stageBundle") { localDir: String, id: String, hash: String ->
            Store(context).stagePending(localDir, id, hash)
        }

        Function("markBundleHealthy") { Store(context).markHealthy() }

        // Persist a crash-detail string (the offending regex pattern + Hermes
        // error) next to the rollback markers, so the rolled-back NEXT launch can
        // upload it via report-bad. The JS-layer Sentry event often can't flush
        // before the process aborts; this on-disk record survives. Mirrors iOS.
        Function("recordBundleError") { detail: String -> Store(context).recordError(detail) }

        // Mark the active OTA bundle bad so the next getJSBundleFile() reverts to
        // the previous bundle. Called from the JS global fatal handler (paired with
        // reload()) so a not-yet-healthy bundle that throws fatally recovers
        // in-session instead of waiting for the crash-launch counter. Mirrors iOS.
        Function("markBundleBad") { Store(context).markBad() }

        // Read-once: { id, hash } of a bundle rolled back since the last call (or
        // null). JS reports it to the server's report-bad endpoint on boot.
        Function("takeRevertedBundle") { Store(context).takeRevertedBundle() }

        AsyncFunction("reload") { Store(context).requestReload() }
    }
}

/**
 * File-backed pointer store mirroring the iOS `Store`. Maintains JSON pointer
 * files in an app-private dir and drives staging / promote / crash-rollback.
 *
 *   current.json        { id, dir, hash }  — active OTA bundle (absent = embedded)
 *   pending.json        { id, dir, hash }  — staged, promoted on next launch
 *   previous.json       { id, dir, hash }  — prior current, rollback target
 *   boot.json           { id, launchCount } — crash tracker
 *   embedded-hash.json  { id, hash }        — cached embedded-bundle hash
 */
class Store(private val context: Context) {
    private val root: File =
        File(context.filesDir, "app-updater").apply { if (!exists()) mkdirs() }

    private val currentFile = File(root, "current.json")
    private val pendingFile = File(root, "pending.json")
    private val previousFile = File(root, "previous.json")
    private val bootFile = File(root, "boot.json")
    private val revertFile = File(root, "revert.json")
    private val revertedFile = File(root, "reverted.json")
    private val embeddedHashFile = File(root, "embedded-hash.json")
    private val errorFile = File(root, "error.json")

    // Crash-launch rollback thresholds (mirrors iOS). A freshly-promoted bundle is
    // on TRIAL until its first markHealthy: it rolls back after the 2nd un-healthy
    // launch (ONE visible crash, then recovery) — the backstop behind the JS fatal
    // handler's in-session revert. A bundle that was already healthy keeps the more
    // generous threshold so a slow first paint that's force-quit before markHealthy
    // isn't mis-read as a crash.
    private val rollbackAfterTrialLaunches = 2
    private val rollbackAfterLaunches = 3

    fun embeddedId(): String = resString("tinycld_bundle_id") ?: "embedded"

    fun embeddedRuntime(): String = resString("tinycld_runtime_version") ?: ""

    fun currentId(): String? = readJSON(currentFile)?.optString("id")?.ifEmpty { null }

    private fun currentDir(): String? = readJSON(currentFile)?.optString("dir")?.ifEmpty { null }

    private fun currentHash(): String? = readJSON(currentFile)?.optString("hash")?.ifEmpty { null }

    fun stagePending(dir: String, id: String, hash: String) {
        writeJSON(pendingFile, JSONObject().put("id", id).put("dir", dir).put("hash", hash))
    }

    /**
     * Marks the active OTA bundle healthy so crash-rollback won't revert it.
     * Called from JS once the app reaches a stable state — the earlier it runs,
     * the smaller the window in which a healthy bundle could be mis-rolled-back.
     */
    fun markHealthy() {
        // Clearing boot.json also clears the trial flag, so a later post-healthy
        // crash falls under the generous (non-trial) threshold, not the trial one.
        bootFile.delete()
        // Drop any stale recorded crash detail — once a bundle is healthy, a prior
        // fatal's detail must not attach to a future unrelated rollback.
        errorFile.delete()
    }

    /**
     * Records a crash-detail string (the offending regex pattern + Hermes error,
     * built by the JS fatal handler) for the active bundle, to be uploaded on the
     * next (rolled-back) launch. Overwrites any prior record. Mirrors iOS.
     */
    fun recordError(detail: String) {
        writeJSON(errorFile, JSONObject().put("detail", detail).put("id", currentId() ?: ""))
    }

    /**
     * Flags the active OTA bundle for revert on the next getJSBundleFile(). The JS
     * global fatal handler calls this (then reload()) when a not-yet-healthy bundle
     * throws fatally, so recovery happens within the same reload cycle instead of
     * waiting for the crash-launch counter. Records the current id so promote only
     * honors a revert that still targets that bundle. Mirrors iOS markBad().
     */
    fun markBad() {
        val id = currentId() ?: return
        writeJSON(revertFile, JSONObject().put("id", id))
    }

    /**
     * Returns a map { id, hash } of a bundle rolled back since the last call, then
     * clears it (read-once). The recovered bundle calls this on boot and, if
     * non-null, POSTs report-bad so the server stops advertising the bad bundle to
     * the fleet. Returns null when nothing was rolled back. Mirrors iOS.
     */
    fun takeRevertedBundle(): Map<String, String>? {
        val r = readJSON(revertedFile) ?: return null
        revertedFile.delete()
        val id = r.optString("id").ifEmpty { return null }
        // Drain the recorded crash detail (read-once, paired with the revert).
        val detail = readJSON(errorFile)?.optString("detail") ?: ""
        errorFile.delete()
        return mapOf("id" to id, "hash" to r.optString("hash"), "error" to detail)
    }

    /**
     * Hex SHA-256 of the bundle the app is currently running. For a promoted OTA
     * bundle this is the hash recorded at stage time (matches the server's
     * bundle_hash). With no OTA bundle promoted it's the embedded bundle's hash,
     * computed once and cached keyed by the embedded id. Returns "" if the
     * embedded bundle can't be read — the server tolerates an empty hash.
     */
    fun currentBundleHash(embeddedId: String): String =
        currentHash() ?: embeddedBundleHash(embeddedId)

    private fun embeddedBundleHash(embeddedId: String): String {
        readJSON(embeddedHashFile)?.let { cached ->
            if (cached.optString("id") == embeddedId) {
                cached.optString("hash").ifEmpty { null }?.let { return it }
            }
        }
        return try {
            val digest = MessageDigest.getInstance("SHA-256")
            context.assets.open("index.android.bundle").use { input ->
                val buf = ByteArray(1 shl 16)
                while (true) {
                    val n = input.read(buf)
                    if (n < 0) break
                    digest.update(buf, 0, n)
                }
            }
            val hex = digest.digest().joinToString("") { "%02x".format(it) }
            writeJSON(embeddedHashFile, JSONObject().put("id", embeddedId).put("hash", hex))
            hex
        } catch (_: Exception) {
            ""
        }
    }

    /**
     * Streaming lowercase-hex SHA-256 of the file at [path], read in 64KB chunks
     * (mirrors the embedded-hash path) so a large .hbc never lands in memory all
     * at once. Returns null if the file can't be read — the caller treats null as
     * a verification failure (rollback), which is the safe direction.
     */
    private fun sha256HexOfFile(path: String): String? =
        try {
            val digest = MessageDigest.getInstance("SHA-256")
            File(path).inputStream().use { input ->
                val buf = ByteArray(1 shl 16)
                while (true) {
                    val n = input.read(buf)
                    if (n < 0) break
                    digest.update(buf, 0, n)
                }
            }
            digest.digest().joinToString("") { "%02x".format(it) }
        } catch (_: Exception) {
            null
        }

    /**
     * Called from the RN host's `getJSBundleFile()` BEFORE the React instance
     * loads. Promotes a pending bundle, applies crash-rollback, and returns the
     * `.hbc` path to load — or `null` to fall back to the embedded bundle.
     */
    fun resolveBundlePath(): String? {
        promotePendingIfAny()
        // A JS fatal handler may have flagged the current bundle bad (markBad) just
        // before triggering this reload. Honor it FIRST — if the flag still targets
        // the current bundle, revert before loading anything, so an in-session
        // recovery never reloads the bad bundle. A stale flag is cleared + ignored.
        readJSON(revertFile)?.optString("id")?.ifEmpty { null }?.let { revertId ->
            revertFile.delete()
            if (revertId == currentId()) {
                rollbackToPrevious("js-fatal markBad")
                return resolveAfterRollback()
            }
        }
        val dir = currentDir() ?: return null
        val id = currentId() ?: return null
        val bundlePath = locateHbc(dir)
        if (bundlePath == null) {
            rollbackToPrevious("staged .hbc missing")
            return resolveAfterRollback()
        }
        // Defense-in-depth behind the JS-side SHA-256 verify (which runs once,
        // pre-stage). The staged bundle lives in a writable app-private dir and
        // isn't re-checked until the next cold launch, so re-hash the .hbc here
        // and refuse to load it if it no longer matches the hash recorded at stage
        // time. This catches post-stage tampering / corruption — treat a mismatch
        // exactly like a missing bundle and roll back rather than load unverified
        // bytes. (current.json `hash` == the server's bundle_hash == sha256 of
        // this .hbc; see downloadAndStage + app_native_export.go sha256OfFile.)
        val want = currentHash()
        if (!want.isNullOrEmpty() && !sha256HexOfFile(bundlePath).equals(want, ignoreCase = true)) {
            rollbackToPrevious("hash mismatch")
            return resolveAfterRollback()
        }
        var boot = readJSON(bootFile)
        if (boot == null || boot.optString("id") != id) {
            boot = JSONObject().put("id", id).put("launchCount", 0).put("trial", true)
        }
        val count = boot.optInt("launchCount", 0) + 1
        // A bundle that has never been marked healthy is on trial → roll back after
        // one un-healthy crash; an already-healthy bundle keeps the generous margin.
        val threshold = if (boot.optBoolean("trial", false)) rollbackAfterTrialLaunches else rollbackAfterLaunches
        if (count >= threshold) {
            rollbackToPrevious("crash-launch counter tripped")
            return resolveAfterRollback()
        }
        boot.put("launchCount", count)
        writeJSON(bootFile, boot)
        return bundlePath
    }

    private fun promotePendingIfAny() {
        val pending = readJSON(pendingFile) ?: return
        if (!pending.has("id") || !pending.has("dir")) return
        val pendingId = pending.optString("id")
        // Idempotency guard for the crash window — see the iOS module for the full
        // rationale. This runs before the bridge loads; a kill after `current` is
        // written but before `pending` is deleted would otherwise re-run promote on
        // the next boot and clobber `previous` (the rollback target). If pending
        // already equals current, the promote completed: clear pending and return.
        if (pendingId == readJSON(currentFile)?.optString("id")) {
            pendingFile.delete()
            return
        }
        readJSON(currentFile)?.let { writeJSON(previousFile, it) }
        writeJSON(currentFile, pending)
        pendingFile.delete()
        // Fresh promote: reset the crash tracker AS A TRIAL (rolls back after one
        // un-healthy crash) and clear any stale revert flag from a prior bundle.
        writeJSON(bootFile, JSONObject().put("id", pending.optString("id")).put("launchCount", 0).put("trial", true))
        revertFile.delete()
    }

    /**
     * Roll back to the previous bundle (or embedded if none). Deleting
     * previous.json is critical: without it a rollback target that ALSO crashes
     * twice would loop forever instead of falling through to embedded.
     */
    /**
     * Rolls the active OTA bundle back to the previous one. `reason` names which
     * path fired so the report-bad upload always carries WHY — most rollbacks
     * happen here at native boot, BEFORE any JS runs, so without this the detail
     * field is empty and the server only learns "a bundle rolled back". Mirrors iOS.
     */
    private fun rollbackToPrevious(reason: String) {
        // Record the rolled-back bundle (id + hash) so the recovered bundle can
        // report it bad to the server on the next boot — see takeRevertedBundle().
        readJSON(currentFile)?.let {
            val curId = it.optString("id")
            writeJSON(revertedFile, JSONObject().put("id", curId).put("hash", it.optString("hash")))
            // Ensure error.json carries a detail for the upload. Prefer a richer
            // JS-supplied detail already recorded for THIS bundle (regex pattern +
            // Hermes message); otherwise stamp the native rollback reason.
            val existing = readJSON(errorFile)
            val jsDetail = existing?.optString("detail") ?: ""
            val jsId = existing?.optString("id") ?: ""
            if (jsDetail.isEmpty() || (jsId.isNotEmpty() && jsId != curId)) {
                val launches = readJSON(bootFile)?.optInt("launchCount") ?: 0
                writeJSON(
                    errorFile,
                    JSONObject()
                        .put("detail", "native rollback: $reason (launches=$launches)")
                        .put("id", curId)
                )
            }
        }
        val prev = readJSON(previousFile)
        if (prev != null) writeJSON(currentFile, prev) else currentFile.delete()
        previousFile.delete()
        bootFile.delete()
        revertFile.delete()
    }

    private fun resolveAfterRollback(): String? {
        val dir = currentDir() ?: return null
        return locateHbc(dir)
    }

    fun requestReload() {
        UiThreadUtil.runOnUiThread {
            val app = context.applicationContext as? ReactApplication
            app?.reactHost?.reload("app-updater OTA reload")
        }
    }

    // `dir` is a filesystem path. JS stages from expo-file-system, whose
    // documentDirectory is a `file://` URI; strip the scheme defensively so a
    // file:// prefix is never treated as a literal path component (which would
    // make the .hbc unfindable and silently roll back to embedded).
    private fun locateHbc(dir: String): String? =
        File(File(dir.removePrefix("file://")), "native/android")
            .walkTopDown()
            .firstOrNull { it.isFile && it.extension == "hbc" }
            ?.path

    private fun resString(name: String): String? {
        val id = context.resources.getIdentifier(name, "string", context.packageName)
        if (id == 0) return null
        return context.getString(id).ifEmpty { null }
    }

    private fun readJSON(file: File): JSONObject? {
        if (!file.exists()) return null
        return try {
            JSONObject(file.readText())
        } catch (_: Exception) {
            null
        }
    }

    private fun writeJSON(file: File, value: JSONObject) {
        try {
            // Write to a temp file then atomically rename, so a crash/kill
            // mid-write can't leave a truncated pointer file (which readJSON would
            // see as corrupt → lose the active/rollback pointer). Mirrors the iOS
            // store's `.atomic` write. rename within the same dir is atomic.
            val tmp = File(file.parentFile, "${file.name}.tmp")
            tmp.writeText(value.toString())
            if (!tmp.renameTo(file)) {
                // Fallback if rename fails (e.g. existing target on some FS): a
                // direct write is still better than dropping the update entirely.
                file.writeText(value.toString())
                tmp.delete()
            }
        } catch (_: Exception) {
            // best-effort persistence; a failed write simply leaves prior state
        }
    }
}

File: modules/app-updater/ios/AppUpdaterModule.swift
import CryptoKit
import ExpoModulesCore
import Foundation

public class AppUpdaterModule: Module {
    public func definition() -> ModuleDefinition {
        Name("AppUpdaterModule")

        Function("getEmbeddedId") { embeddedId() }
        Function("getRuntimeVersion") { embeddedRuntimeVersion() }
        Function("getCurrentBundleId") { Store.shared.currentId() ?? embeddedId() }
        Function("getCurrentBundleHash") { Store.shared.currentBundleHash(embeddedId: embeddedId()) }

        // Point the store at a server (hex digest of its normalized origin).
        // Everything below then reads and writes that server's own bundle
        // state. JS calls this as soon as an address resolves, so the value is
        // already on disk for the next launch — the loader runs before the
        // bridge and cannot ask JS which server is active.
        Function("setActiveServer") { (serverKey: String) in
            Store.shared.setActiveServer(serverKey)
        }
        Function("getActiveServer") { Store.shared.activeServerKey() ?? "" }

        AsyncFunction("stageBundle") { (localDir: String, id: String, hash: String) in
            try Store.shared.stagePending(dir: localDir, id: id, hash: hash)
        }

        Function("markBundleHealthy") { Store.shared.markHealthy() }

        // Persist a crash-detail string (e.g. the offending regex pattern + the
        // Hermes error) next to the rollback markers, so the rolled-back NEXT
        // launch can upload it via report-bad. The JS-layer Sentry event often
        // can't flush before the process aborts; this on-disk record survives.
        Function("recordBundleError") { (detail: String) in Store.shared.recordError(detail) }

        // Mark the active OTA bundle bad so the NEXT bundleURL() reverts to the
        // previous bundle. Called from the JS global fatal handler when a
        // not-yet-healthy bundle throws fatally — paired with reload() it recovers
        // in-session instead of waiting for the crash-launch counter to trip.
        Function("markBundleBad") { Store.shared.markBad() }

        // Read-once: returns { id, hash } of a bundle rolled back since the last
        // call (or nil). JS reports it to the server's report-bad endpoint on boot.
        Function("takeRevertedBundle") { Store.shared.takeRevertedBundle() }

        AsyncFunction("reload") {
            DispatchQueue.main.async { Store.shared.requestReload() }
        }
    }

    private func embeddedId() -> String {
        Bundle.main.object(forInfoDictionaryKey: "TinyCldBundleId") as? String ?? "embedded"
    }
    private func embeddedRuntimeVersion() -> String {
        Bundle.main.object(forInfoDictionaryKey: "TinyCldRuntimeVersion") as? String ?? ""
    }
}

final class Store {
    static let shared = Store()
    private let fm = FileManager.default

    private var root: URL {
        let base = fm.urls(for: .applicationSupportDirectory, in: .userDomainMask)[0]
            .appendingPathComponent("app-updater", isDirectory: true)
        try? fm.createDirectory(at: base, withIntermediateDirectories: true)
        return base
    }

    /// Names which server's bundle to load. Written by JS the moment a server
    /// address resolves; read here at launch, BEFORE the React bridge exists —
    /// which is precisely why it must be persisted rather than asked for.
    private var activeURL: URL { root.appendingPathComponent("active.json") }

    /// Per-server pointer state. Each server the user connects to keeps its own
    /// bundle and its own crash-tracking, because in a multi-org deployment
    /// every org runs a DIFFERENT build with a different package set. With one
    /// shared slot, switching orgs made each foreground see the other org's
    /// bundle as "not current" and re-download it — a thrash loop, with a full
    /// JS reload each time.
    ///
    /// `serverKey` is a hex digest of the normalized origin, computed JS-side
    /// (see core/lib/app-updater/server-key.ts). It is validated here as pure
    /// hex before being used as a path component, so no origin can escape this
    /// directory or collide with another.
    private func serverRoot(_ serverKey: String) -> URL {
        root.appendingPathComponent("servers", isDirectory: true)
            .appendingPathComponent(serverKey, isDirectory: true)
    }

    /// The active server's directory, created on demand. Falls back to a
    /// "legacy" bucket when nothing is active yet so the store is never
    /// rootless; that bucket is empty on a fresh install, which resolves to the
    /// embedded bundle — the correct answer.
    private var scope: URL {
        let dir = serverRoot(activeServerKey() ?? "legacy")
        try? fm.createDirectory(at: dir, withIntermediateDirectories: true)
        return dir
    }

    /// Reads the active server key, rejecting anything that is not a plain hex
    /// digest. A malformed value is treated as absent rather than joined into a
    /// path.
    func activeServerKey() -> String? {
        guard let k = readJSON(activeURL)?["serverKey"] as? String, isValidServerKey(k) else {
            return nil
        }
        return k
    }

    private func isValidServerKey(_ k: String) -> Bool {
        !k.isEmpty && k.count <= 64 && k.allSatisfy { $0.isHexDigit }
    }

    /// Points the store at `serverKey`. Idempotent: re-writing the same key is
    /// a no-op, so the common case (every foreground re-asserts the address)
    /// does not churn the file the launch path depends on.
    func setActiveServer(_ serverKey: String) {
        guard isValidServerKey(serverKey) else { return }
        if activeServerKey() == serverKey { return }
        writeJSON(activeURL, ["serverKey": serverKey])
    }

    private var currentURL: URL { scope.appendingPathComponent("current.json") }
    private var pendingURL: URL { scope.appendingPathComponent("pending.json") }
    private var previousURL: URL { scope.appendingPathComponent("previous.json") }
    private var bootURL: URL { scope.appendingPathComponent("boot.json") }
    private var revertURL: URL { scope.appendingPathComponent("revert.json") }
    private var revertedURL: URL { scope.appendingPathComponent("reverted.json") }
    private var errorURL: URL { scope.appendingPathComponent("error.json") }

    // The embedded bundle belongs to the BINARY, not to any server, so its
    // cached hash stays at the root and is shared across every server scope.
    private var embeddedHashURL: URL { root.appendingPathComponent("embedded-hash.json") }

    // Crash-launch rollback thresholds: a promoted OTA bundle is rolled back after
    // this many launches that never reach the JS "healthy" signal.
    //
    // A freshly-promoted bundle is on TRIAL until its first markHealthy: it rolls
    // back after the 2nd un-healthy launch (i.e. ONE visible crash, then recovery
    // on the next launch). This is the "roll back on first crash" backstop behind
    // the JS fatal handler's in-session revert.
    //
    // A bundle that was ALREADY healthy at least once keeps the more generous
    // threshold so a normal boot that's force-quit before markHealthy (slow first
    // paint) isn't mis-read as a crash and rolled back.
    private let rollbackAfterTrialLaunches = 2
    private let rollbackAfterLaunches = 3

    /// Reads a string field from current.json, treating an empty string as nil —
    /// matching the Android store's `optString(...).ifEmpty { null }`, so an empty
    /// pointer field is uniformly "absent" across platforms.
    private func currentString(_ key: String) -> String? {
        guard let v = readJSON(currentURL)?[key] as? String, !v.isEmpty else { return nil }
        return v
    }
    func currentId() -> String? { currentString("id") }
    private func currentDir() -> String? { currentString("dir") }
    private func currentHash() -> String? { currentString("hash") }

    func stagePending(dir: String, id: String, hash: String) throws {
        writeJSON(pendingURL, ["id": id, "dir": dir, "hash": hash])
    }

    /// Marks the active OTA bundle healthy so crash-rollback won't revert it.
    /// Called from JS after the app reaches a stable state — the earlier it runs,
    /// the smaller the window in which a healthy bundle could be mis-rolled-back.
    /// Removing boot.json also clears the trial flag, so a later post-healthy crash
    /// falls under the generous (non-trial) threshold rather than the trial one.
    /// Also drop any stale recorded crash detail — once a bundle is healthy, a
    /// prior fatal's detail must not attach to a future unrelated rollback.
    func markHealthy() {
        try? fm.removeItem(at: bootURL)
        try? fm.removeItem(at: errorURL)
    }

    /// Flags the active OTA bundle for revert on the next bundleURL(). The JS
    /// global fatal handler calls this (then reload()) when a not-yet-healthy
    /// bundle throws fatally, so recovery happens within the same reload cycle
    /// instead of waiting for the crash-launch counter. Records the current id so
    /// promotePendingIfAny only honors a revert that still targets that bundle.
    func markBad() {
        guard let id = currentId() else { return }
        writeJSON(revertURL, ["id": id])
    }

    /// Records a crash-detail string (the offending regex pattern + Hermes error,
    /// built by the JS fatal handler) for the active bundle, to be uploaded on the
    /// next (rolled-back) launch. Overwrites any prior record — the most recent
    /// fatal is the relevant one. Best-effort; a write failure is non-fatal.
    func recordError(_ detail: String) {
        writeJSON(errorURL, ["detail": detail, "id": currentId() ?? ""])
    }

    /// Returns `[id, hash, error, serverKey]` of a bundle that was rolled back
    /// since the last call, then clears it (read-once). The recovered bundle
    /// calls this on boot and, if non-nil, POSTs report-bad so the server stops
    /// advertising the bad bundle to the fleet. `error` carries the persisted
    /// crash detail (e.g. the regex pattern that aborted Hermes) when one was
    /// recorded, else "". Returns nil when nothing was rolled back.
    ///
    /// `serverKey` names the server the bad bundle CAME FROM. Without it the
    /// report would go to whichever server happens to be active now, which
    /// after a switch is the wrong one — it would blocklist a bundle id on a
    /// server that never served it while the actually-bad bundle kept being
    /// advertised by its own.
    func takeRevertedBundle() -> [String: String]? {
        // Scan every server's scope, not just the active one: the rollback
        // happens at native boot, and JS may resolve a different address before
        // it drains the marker.
        let serversRoot = root.appendingPathComponent("servers", isDirectory: true)
        let keys = (try? fm.contentsOfDirectory(atPath: serversRoot.path)) ?? []
        for key in keys + ["legacy"] {
            let dir = serverRoot(key)
            let revertedFile = dir.appendingPathComponent("reverted.json")
            guard let r = readJSON(revertedFile) else { continue }
            try? fm.removeItem(at: revertedFile)
            let id = r["id"] as? String ?? ""
            if id.isEmpty { continue }
            // Drain the recorded crash detail (read-once, paired with the revert).
            let errorFile = dir.appendingPathComponent("error.json")
            let detail = (readJSON(errorFile)?["detail"] as? String) ?? ""
            try? fm.removeItem(at: errorFile)
            return [
                "id": id,
                "hash": r["hash"] as? String ?? "",
                "error": detail,
                "serverKey": key == "legacy" ? "" : key,
            ]
        }
        return nil
    }

    /// Hex SHA-256 of the bundle the app is currently running. For a promoted OTA
    /// bundle this is the hash recorded at stage time (matches the server's
    /// bundle_hash). With no OTA bundle promoted it's the embedded bundle's hash,
    /// computed once and cached keyed by the embedded id so a later app version
    /// (new embedded id) re-hashes. Returns "" if the embedded bundle can't be
    /// found/hashed — the server tolerates an empty hash (falls back to id check).
    func currentBundleHash(embeddedId: String) -> String {
        if let h = currentHash() { return h }
        return embeddedBundleHash(embeddedId: embeddedId)
    }

    private func embeddedBundleHash(embeddedId: String) -> String {
        if let cached = readJSON(embeddedHashURL),
            cached["id"] as? String == embeddedId,
            let h = cached["hash"] as? String {
            return h
        }
        guard let url = Bundle.main.url(forResource: "main", withExtension: "jsbundle"),
            let data = try? Data(contentsOf: url) else {
            return ""
        }
        let hex = SHA256.hash(data: data).map { String(format: "%02x", $0) }.joined()
        writeJSON(embeddedHashURL, ["id": embeddedId, "hash": hex])
        return hex
    }

    /// Streaming lowercase-hex SHA-256 of the file at `path`, read in 64KB chunks
    /// so a large .hbc never lands in memory all at once. Returns nil if the file
    /// can't be opened/read — the caller treats nil as a verification failure
    /// (rollback), which is the safe direction.
    private func sha256HexOfFile(at path: String) -> String? {
        guard let handle = FileHandle(forReadingAtPath: path) else { return nil }
        defer { try? handle.close() }
        var hasher = SHA256()
        while true {
            let chunk: Data
            do {
                guard let next = try handle.read(upToCount: 1 << 16) else { break }
                chunk = next
            } catch {
                return nil
            }
            if chunk.isEmpty { break }
            hasher.update(data: chunk)
        }
        return hasher.finalize().map { String(format: "%02x", $0) }.joined()
    }

    /// Called from the AppDelegate's `bundleURL()` BEFORE the React bridge loads.
    /// Promotes a pending bundle, applies crash-rollback, and returns the `.hbc`
    /// to load — or `nil` to fall back to the embedded `main.jsbundle`.
    func resolveBundleURL() -> URL? {
        promotePendingIfAny()
        // A JS fatal handler may have flagged the current bundle bad (markBad) just
        // before triggering this reload. Honor it FIRST — if the flag still targets
        // the current bundle, revert to previous before doing anything else, so an
        // in-session recovery never even loads the bad bundle again. A stale flag
        // (id no longer current) is cleared and ignored.
        if let revertId = readJSON(revertURL)?["id"] as? String {
            try? fm.removeItem(at: revertURL)
            if revertId == currentId() {
                rollbackToPrevious(reason: "js-fatal markBad")
                return resolveAfterRollback()
            }
        }
        guard let dir = currentDir(), let id = currentId() else { return nil }
        guard let bundlePath = locateHbc(in: dir) else {
            rollbackToPrevious(reason: "staged .hbc missing")
            return resolveAfterRollback()
        }
        // Defense-in-depth behind the JS-side SHA-256 verify (which runs once,
        // pre-stage). The staged bundle lives in a writable app dir and isn't
        // re-checked until the next cold launch, so re-hash the .hbc here and
        // refuse to load it if it no longer matches the hash recorded at stage
        // time. This catches post-stage tampering / corruption — treat a mismatch
        // exactly like a missing bundle and roll back rather than load unverified
        // bytes. (current.json `hash` == the server's bundle_hash == sha256 of
        // this .hbc; see downloadAndStage + app_native_export.go sha256OfFile.)
        if let want = currentHash(), !want.isEmpty,
            sha256HexOfFile(at: bundlePath)?.caseInsensitiveCompare(want) != .orderedSame {
            rollbackToPrevious(reason: "hash mismatch")
            return resolveAfterRollback()
        }
        var boot = readJSON(bootURL) ?? ["id": id, "launchCount": 0, "trial": true]
        if (boot["id"] as? String) != id { boot = ["id": id, "launchCount": 0, "trial": true] }
        let count = (boot["launchCount"] as? Int ?? 0) + 1
        // A bundle that has never been marked healthy is on trial → roll back after
        // one un-healthy crash; an already-healthy bundle keeps the generous margin.
        let threshold = (boot["trial"] as? Bool ?? false) ? rollbackAfterTrialLaunches : rollbackAfterLaunches
        if count >= threshold {
            rollbackToPrevious(reason: "crash-launch counter tripped")
            return resolveAfterRollback()
        }
        boot["launchCount"] = count
        writeJSON(bootURL, boot)
        return URL(fileURLWithPath: bundlePath)
    }

    private func promotePendingIfAny() {
        guard let p = readJSON(pendingURL),
            let pendingId = p["id"] as? String, p["dir"] is String else { return }
        // Idempotency guard for the crash window. This runs before the React bridge
        // loads, so the OS can kill us mid-promote. The three writes below are not
        // atomic as a set and `pending` is removed LAST — without this guard, a kill
        // after `current` is written but before `pending` is removed would, on the
        // next boot, re-run promote and overwrite `previous` (the good rollback
        // target) with the bundle we just promoted, destroying the safety net. If
        // the pending id already equals current, the promote already completed: just
        // clear the leftover pending pointer and return.
        if pendingId == (readJSON(currentURL)?["id"] as? String) {
            try? fm.removeItem(at: pendingURL)
            return
        }
        if let cur = readJSON(currentURL) { writeJSON(previousURL, cur) }
        writeJSON(currentURL, p)
        try? fm.removeItem(at: pendingURL)
        // Fresh promote: reset the crash tracker AS A TRIAL (rolls back after one
        // un-healthy crash) and clear any stale revert flag from a prior bundle.
        writeJSON(bootURL, ["id": p["id"] as? String ?? "", "launchCount": 0, "trial": true])
        try? fm.removeItem(at: revertURL)
    }

    /// Rolls the active OTA bundle back to the previous one. `reason` names which
    /// path fired (crash-counter, missing-hbc, hash-mismatch, js-fatal) so the
    /// report-bad upload always carries WHY — most rollbacks happen here at native
    /// boot, BEFORE any JS runs, so without this the detail field is empty and the
    /// server only learns "a bundle rolled back", not why.
    private func rollbackToPrevious(reason: String) {
        // Record the bundle being rolled back (id + hash) so the recovered bundle
        // can report it bad to the server on the next boot — takeRevertedBundle().
        // The crash handler can't report at crash time (the process is dying), so
        // we stash it and let the good bundle POST it once it's safely running.
        if let cur = readJSON(currentURL) {
            let curId = cur["id"] as? String ?? ""
            writeJSON(revertedURL, ["id": curId, "hash": cur["hash"] as? String ?? ""])
            // Ensure error.json carries a detail for the upload. Prefer a richer
            // JS-supplied detail already recorded for THIS bundle (the regex
            // pattern + Hermes message from the fatal handler); otherwise stamp the
            // native rollback reason. Either way the report-bad row is meaningful.
            let existing = readJSON(errorURL)
            let jsDetail = existing?["detail"] as? String
            let jsId = existing?["id"] as? String
            if jsDetail == nil || jsDetail!.isEmpty || (jsId != nil && jsId != curId) {
                let boot = readJSON(bootURL)
                let launches = boot?["launchCount"] as? Int ?? 0
                writeJSON(
                    errorURL,
                    ["detail": "native rollback: \(reason) (launches=\(launches))", "id": curId]
                )
            }
        }
        if let prev = readJSON(previousURL) { writeJSON(currentURL, prev) }
        else { try? fm.removeItem(at: currentURL) }
        try? fm.removeItem(at: previousURL)
        try? fm.removeItem(at: bootURL)
        try? fm.removeItem(at: revertURL)
    }
    private func resolveAfterRollback() -> URL? {
        guard let dir = currentDir() else { return nil }
        return locateHbc(in: dir).map { URL(fileURLWithPath: $0) }
    }

    func requestReload() {
        AppUpdaterReload.trigger(withReason: "app-updater OTA reload")
    }

    private func locateHbc(in dir: String) -> String? {
        // `dir` is a filesystem path. JS stages from expo-file-system, whose
        // documentDirectory is a `file://` URI; URL(fileURLWithPath:) would treat
        // that prefix as a literal path component and never find the .hbc
        // (silently rolling back to embedded). Resolve a file:// URI properly,
        // otherwise treat the string as a plain path.
        let dirURL = dir.hasPrefix("file://")
            ? (URL(string: dir) ?? URL(fileURLWithPath: dir))
            : URL(fileURLWithPath: dir)
        let base = dirURL.appendingPathComponent("native/ios", isDirectory: true)
        guard let en = fm.enumerator(at: base, includingPropertiesForKeys: nil) else { return nil }
        for case let f as URL in en where f.pathExtension == "hbc" { return f.path }
        return nil
    }
    private func readJSON(_ u: URL) -> [String: Any]? {
        guard let d = try? Data(contentsOf: u) else { return nil }
        return (try? JSONSerialization.jsonObject(with: d)) as? [String: Any]
    }
    private func writeJSON(_ u: URL, _ v: [String: Any]) {
        if let d = try? JSONSerialization.data(withJSONObject: v) { try? d.write(to: u, options: .atomic) }
    }
}
